docs/yum.8
.\" yum - Yellowdog Updater Modified
.TH "yum" "8" ""  "Seth Vidal" ""
.SH "NAME"
yum \- Yellowdog Updater Modified
.SH "SYNOPSIS"
\fByum\fP [options] [command] [package ...]
.SH "DESCRIPTION"
.PP 
\fByum\fP is an interactive, rpm based, package manager. It can automatically
perform system updates, including dependency analysis and obsolete processing
based on "repository" metadata. It can also perform installation of new
packages, removal of old packages and perform queries on the installed and/or
available packages among many other commands/services (see below)\&. \fByum\fP
is similar to other high level package managers like apt\-get and smart\&.
.PP
While there are some graphical interfaces directly to the \fByum\fP code, more
recent graphical interface development is happening with PackageKit and the
gnome\-packagekit application\&.
.PP 
\fIcommand\fP is one of:
.br 
.I \fR * install package1 [package2] [\&.\&.\&.]
.br 
.I \fR * update [package1] [package2] [\&.\&.\&.]
.br 
.I \fR * update-to [package1] [package2] [\&.\&.\&.]
.br 
.I \fR * check\-update
.br 
.I \fR * upgrade [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * upgrade-to [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * distribution-synchronization [package1] [package2] [\&.\&.\&.] 
.br
.I \fR * remove | erase package1 [package2] [\&.\&.\&.]
.br 
.I \fR * list [\&.\&.\&.]
.br 
.I \fR * info [\&.\&.\&.]
.br 
.I \fR * provides  | whatprovides feature1 [feature2] [\&.\&.\&.]
.br  
.I \fR * clean [ packages | metadata | expire-cache | rpmdb | plugins | all ]
.br
.I \fR * makecache
.br
.I \fR * groups  [\&.\&.\&.]
.br
.I \fR * search string1 [string2] [\&.\&.\&.]
.br
.I \fR * shell [filename]
.br
.I \fR * resolvedep dep1 [dep2] [\&.\&.\&.] 
    (maintained for legacy reasons only - use repoquery or yum provides)
.br
.I \fR * localinstall rpmfile1 [rpmfile2] [\&.\&.\&.] 
    (maintained for legacy reasons only - use install)
.br
.I \fR * localupdate rpmfile1 [rpmfile2] [\&.\&.\&.]
    (maintained for legacy reasons only - use update)
.br
.I \fR * reinstall package1 [package2] [\&.\&.\&.] 
.br
.I \fR * downgrade package1 [package2] [\&.\&.\&.] 
.br
.I \fR * deplist package1 [package2] [\&.\&.\&.] 
.br
.I \fR * repolist [all|enabled|disabled] 
.br
.I \fR * version [ all | installed | available | group-* | nogroups* | grouplist | groupinfo ]
.br
.I \fR * history [info|list|packages-list|packages-info|summary|addon-info|redo|undo|rollback|new|sync|stats] 
.br
.I \fR * load-transaction [txfile]
.br
.I \fR * check
.br 
.I \fR * help [command] 
.br
.PP 
Unless the \-\-help or \-h option is given, one of the above commands
must be present\&.
.PP
Repository configuration is honored in all operations.
.PP 
.IP "\fBinstall\fP"
Is used to install the latest version of a package or
group of packages while ensuring that all dependencies are
satisfied\&.  (See \fBSpecifying package names\fP for more information) 
If no package matches the given package name(s), they are assumed to be a shell 
glob and any matches are then installed\&. If the name starts with an 
@ character the rest of the name is used as though passed to the groupinstall
command\&. If the name starts with a - character, then a search is done within
the transaction and any matches are removed. If the name is a file, then install works
like localinstall\&. If the name doesn't match a package, then package
"provides" are searched (e.g. "_sqlitecache.so()(64bit)") as are
filelists (Eg. "/usr/bin/yum"). Also note that for filelists, wildcards will
match multiple packages\&.
.IP 
.IP "\fBupdate\fP"
If run without any packages, update will update every currently
installed package.  If one or more packages or package globs are specified, Yum will
only update the listed packages\&.  While updating packages, \fByum\fP
will ensure that all dependencies are satisfied\&. (See \fBSpecifying package names\fP for more information) 
If the packages or globs specified match to packages which are not currently installed then update will
not install them\&. update operates on groups, files, provides and filelists
just like the "install" command\&.

If the main obsoletes configure option is true (default) or the \-\-obsoletes
flag is present \fByum\fP will include package 
obsoletes in its calculations - this makes it better for distro\-version 
changes, for example: upgrading from somelinux 8.0 to somelinux 9.

Note that "\fBupdate\fP" works on installed packages first, and only if there
are no matches does it look for available packages. The difference is most
noticeable when you do "\fBupdate\fP foo-1-2" which will act exactly as
"\fBupdate\fP foo" if foo-1-2 is installed. You can use the "\fBupdate-to\fP"
if you'd prefer that nothing happen in the above case.
.IP 
.IP "\fBupdate-to\fP"
This command works like "\fBupdate\fP" but always specifies the version of the
package we want to update to.
.IP 
.IP "\fBcheck\-update\fP"
Implemented so you could know if your machine had any updates that needed to
be applied without running it interactively. Returns exit value of 100 if
there are packages available for an update. Also returns a list of the packages
to be updated in list format. Returns 0 if no packages are available for
update. Returns 1 if an error occurred.
Running in verbose mode also shows obsoletes.
.IP
.IP "\fBupgrade\fP"
Is the same as the update command with the \-\-obsoletes flag set. See update 
for more details.
.IP 
.IP "\fBupgrade-to\fP"
This command works like "\fBupgrade\fP" but always specifies the version of the
package we want to update to.
.IP 
.IP "\fBdistribution\-synchronization\fP or \fBdistro\-sync\fP"
Synchronizes the installed package set with the latest packages available, this
is done by either obsoleting, upgrading or downgrading as appropriate. This will
"normally" do the same thing as the upgrade command however if you have the
package FOO installed at version 4, and the latest available is only
version 3, then this command will \fBdowngrade\fP FOO to version 3.

If you give the optional argument "full", then the command will also reinstall
packages where the install checksum and the available checksum do not match. And
remove old packages (can be used to sync. rpmdb versions). The optional argument
"different" can be used to specify the default operation.

This command does not perform operations on groups, local packages or negative
selections.
.IP 
.IP "\fBremove\fP or \fBerase\fP"
Are used to remove the specified packages from the system
as well as removing any packages which depend on the package being
removed\&. remove operates on groups, files, provides and filelists just like
the "install" command\&.(See \fBSpecifying package names\fP for more information) 

Note that "yum" is included in the protected_packages configuration, by default.
So you can't accidentally remove yum itself.
.IP 
.IP "\fBlist\fP"
Is used to list various information about available
packages; more complete details are available in the \fIList Options\fP
section below\&.
.IP 
.IP "\fBprovides\fP or \fBwhatprovides\fP"
Is used to find out which package provides some feature
or file. Just use a specific name or a file-glob-syntax wildcards to list
the packages available or installed that provide that feature or file\&.
.IP 
.IP "\fBsearch\fP"
This is used to find packages when you know something about the package but
aren't sure of it's name. By default search will try searching just package
names and summaries, but if that "fails" it will then try descriptions and url.

Yum search orders the results so that those packages matching more terms will
appear first.

You can force searching everything by specifying "all" as the first argument.
.IP 
.IP "\fBinfo\fP"
Is used to list a description and summary information about available
packages; takes the same arguments as in the \fIList Options\fP
section below\&.
.IP 
.IP "\fBclean\fP"
Is used to clean up various things which accumulate in the
\fByum\fP cache directory over time.  More complete details can be found in
the \fIClean Options\fP section below\&.
.IP 
.IP "\fBmakecache\fP"
Is used to download and make usable all the metadata for the currently enabled
\fByum\fP repos.
.IP 
.IP "\fBgroups\fP"
A command, new in 3.4.2, that collects all the subcommands that act on groups together.

"\fBgroup install\fP" is used to install all of the individual packages in a group, of the specified
types (this works as if you'd taken each of those package names and put them on
the command line for a "yum install" command).
 The group_package_types configuration option specifies which types will
be installed.

"\fBgroup update\fP" is just an alias for groupinstall, which will do the right thing because
"yum install X" and "yum update X" do the same thing, when X is already
installed.

"\fBgroup list\fP" is used to list the available groups from all \fByum\fP repos. Groups are marked
as "installed" if all mandatory packages are installed, or if a group doesn't
have any mandatory packages then it is installed if any of the optional or
default package are installed.
The optional "hidden" argument will also list groups marked as not being
"user visible". If you pass the \-v option, to enable verbose mode, then the
groupids are displayed.

"\fBgroup remove\fP" is used to remove all of the packages in a group, unlike "groupinstall" this
will remove everything regardless of group_package_types. It is worth pointing
out that packages can be in more than one group, so "group install X Y" followed
by "group remove Y" does not do give you the same result as "group install X".

The groupremove_leaf_only configuration changes the behaviour of this command
to only remove packages which aren't required by something else.

"\fBgroup info\fP" is used to give the description and package list of a group (and which type
those packages are marked as). Note that you can use the yum-filter-data and
yum-list-data plugins to get/use the data the other way around (i.e. what
groups own packages need updating). If you pass the \-v option, to enable verbose
mode, then the package names are matched against installed/available packages
similar to the list command.
.IP
.IP "\fBshell\fP"
Is used to enter the 'yum shell', when a filename is specified the contents of
that file is executed in yum shell mode. See \fIyum-shell(8)\fP for more info
.IP
.IP "\fBresolvedep\fP"
Is used to list packages providing the specified dependencies, at most one
package is listed per dependency. This command is maintained for legacy
reasons only, use repoquery instead.
.IP
.IP "\fBlocalinstall\fP"
Is used to install a set of local rpm files. If required the enabled 
repositories will be used to resolve dependencies. Note that the install command
will do a local install, if given a filename. This command is maintained for legacy
reasons only.
.IP
.IP "\fBlocalupdate\fP"
Is used to update the system by specifying local rpm files. Only the specified 
rpm files of which an older version is already installed will be installed,
the remaining specified packages will be ignored.
If required the enabled repositories will be used to resolve dependencies. Note
that the update command will do a local update, if given a filename. This command is maintained for
legacy reasons only.
.IP
.IP "\fBreinstall\fP"
Will reinstall the identically versioned package as is currently installed. 
This does not work for "installonly" packages, like Kernels. reinstall operates
on groups, files, provides and filelists just like the "install" command\&.
.IP
.IP "\fBdowngrade\fP"
Will try and downgrade a package from the version currently installed to the
previously highest version (or the specified version).
The depsolver will not necessarily work, but if you specify all the packages it
should work (thus, all the simple cases will work). Also this does not
work for "installonly" packages, like Kernels. downgrade operates
on groups, files, provides, filelists and rpm files just like the "install" command\&.
.IP
.IP "\fBdeplist\fP"
Produces a list of all dependencies and what packages provide those
dependencies for the given packages. As of 3.2.30 it now just shows the latest
version of each package that matches (this can be changed by
using --showduplicates) and it only shows the newest providers (which can be
changed by using --verbose).
.IP
.IP "\fBrepolist\fP"
Produces a list of configured repositories. The default is to list all
enabled repositories. If you pass \-v, for verbose mode, more information is
listed. If the first argument is 'enabled', 'disabled' or 'all' then the command
will list those types of repos.

You can pass repo id or name arguments, or wildcards which to match against
both of those. However if the id or name matches exactly then the repo will
be listed even if you are listing enabled repos. and it is disabled.

In non-verbose mode the first column will start with a '*' if the repo. has
metalink data and the latest metadata is not local. For non-verbose mode the
last column will also display the number of packages in the repo. and (if there
are any user specified excludes) the number of packages excluded.

One last special feature of repolist, is that if you are in non-verbose mode
then yum will ignore any repo errors and output the information it can get
(Eg. "yum clean all; yum -C repolist" will output something, although the
package counts/etc. will be zeroed out).
.IP
.IP "\fBversion\fP"
Produces a "version" of the rpmdb, and of the enabled repositories if "all" is
given as the first argument. You can also specify version groups in the
version-groups configuration file. If you pass \-v, for verbose mode, more
information is listed. The version is calculated by taking an SHA1 hash of the
packages (in sorted order), and the checksum_type/checksum_data entries from
the yumdb. Note that this rpmdb version is now also used significantly within
yum (esp. in yum history).

The version command will now show "groups" of packages as a separate version,
and so takes sub-commands:

"version grouplist" - List the defined version groups.

"version groupinfo" - Get the complete list of packages within one or more version groups.

"version installed" - This is the default, only show the version information for installed packages.

"version available" - Only show the version information for available packages.

"version all" - Show the version information for installed and available packages.

"version nogroups | nogroups-*" - Just show the main version information.

"version group-*" - Just show the grouped version information, if more arguments are given then only show the data for those groups.

.IP
.IP "\fBhistory\fP"
The history command allows the user to view what has happened in past
transactions (assuming the history_record config. option is set). You can use
info/list/packages-list/packages-info/summary to view what happened,
undo/redo/rollback to act on that information and new to start a new history
file.

The info/list/summary commands take either a transaction id or a package (with
wildcards, as in \fBSpecifying package names\fP), all three can also be passed
no arguments. list can be passed the keyword "all" to list all the transactions.

The packages-list/packages-info commands takes a package  (with wildcards, as in
\fBSpecifying package names\fP). And show data from the point of view of that
package.

The undo/redo/rollback commands take either a single transaction id or the
keyword last and an offset from the last transaction (Eg. if you've done 250
transactions, "last" refers to transaction 250, and "last-4" refers to
transaction 246).
The redo command can also take some optional arguments before you specify the
transaction. "force-reinstall" tells it reinstall any packages that were
installed in that transaction (via. install, upgrade or downgrade).
"force-remove" tells it to forcibly remove any packages that were updated or
downgraded.

The undo/redo commands act on the specified transaction, undo'ing or repeating
the work of that transaction. While the rollback command will undo all
transactions up to the point of the specified transaction. For example, if you
have 3 transactions, where package A; B and C where installed respectively.
Then "undo 1" will try to remove package A, "redo 1" will try to install package
A (if it is not still installed), and "rollback 1" will try to remove packages
B and C. Note that after a "rollback 1" you will have a fourth transaction,
although the ending rpmdb version (see: yum version) should be the same in
transactions 1 and 4.

The addon-info command takes a transaction ID, and the packages-list command
takes a package (with wildcards).

The stats command shows some statistics about the current history DB.

The sync commands allows you to change the rpmdb/yumdb data stored for any
installed packages, to whatever is in the current rpmdb/yumdb (this is mostly
useful when this data was not stored when the package went into the history DB).

In "history list" you can change the behaviour of the 2nd column via. the
configuration option history_list_view.

In "history list" output the Altered column also gives some extra information
if there was something not good with the transaction (this is also shown at the
end of the package column in the packages-list command).

.I \fB>\fR - The rpmdb was changed, outside yum, after the transaction.
.br
.I \fB<\fR - The rpmdb was changed, outside yum, before the transaction.
.br
.I \fB*\fR - The transaction aborted before completion.
.br
.I \fB#\fR - The transaction completed, but with a non-zero status.
.br
.I \fBE\fR - The transaction completed fine, but had warning/error output during the transaction.
.br
.I \fBP\fR - The transaction completed fine, but problems already existed in the rpmdb.
.br
.I \fBs\fR - The transaction completed fine, but --skip-broken was enabled and had to skip some packages.
.br


.IP
.IP "\fBload-transaction\fP"
This command will re-load a saved yum transaction file, this allows you to
run a transaction on one machine and then use it on another.
The two common ways to get a saved yum transaction file are from
"yum -q history addon-info last saved_tx" or via. the automatic saves in
$TMPDIR/yum_save_tx.* when a transaction is solved but not run.

.IP
.IP "\fBcheck\fP"
Checks the local rpmdb and produces information on any problems it finds. You
can pass the check command the arguments "dependencies" or "duplicates", to
limit the checking that is performed (the default is "all" which does both).

The info command can also take ranges of transaction ids, of the form
start..end, which will then display a merged history as if all the
transactions in the range had happened at once\&.
.br
Eg. "history info 1..4" will merge the first four transactions and display them
as a single transaction.
.IP
.IP "\fBhelp\fP"
Produces help, either for all commands or if given a command name then the help
for that particular command\&.
.IP
.PP
.SH "GENERAL OPTIONS"
Most command line options can be set using the configuration file as
well and the descriptions indicate the necessary configuration option
to set\&.
.PP 
.IP "\fB\-h, \-\-help\fP"
Help; display a help message and then quit\&.
.IP "\fB\-y, \-\-assumeyes\fP"
Assume yes; assume that the answer to any question which would be asked 
is yes\&.
.br
Configuration Option: \fBassumeyes\fP
.IP "\fB\-\-assumeno\fP"
Assume no; assume that the answer to any question which would be asked 
is no\&. This option overrides assumeyes, but is still subject to alwaysprompt.
.br
Configuration Option: \fBassumeno\fP

ChangeLog
2011-06-28  James Antill <james@and.org>

	* yum/__init__.py: Bump version to 3.4.3

2011-06-28  James Antill <james@and.org>

	* yum/__init__.py:  Don't call repo.doSetup() if we've already setup the repos. BZ
	 717163.   Test: yum in @core == has    setup.        yum rm @core
	 == hasn't setup.

2011-06-28  James Antill <james@and.org>

	* yum/history.py, yum/rpmtrans.py:  Disable the rpmtrans chroot workaround. Use exclusive locking in
	 history.   The only downside here appears to be that we get
	 .sqlite-journal files in /var/lib/yum/history ... which is probably
	 fine.  Also tested fine as a normal user, with read-only access.

2011-06-27  James Antill <james@and.org>

	* yummain.py: Change the logging for NFS path problems from debug to
	critical.

2011-06-24  Nick Jacek <njacek@redhat.com>

	* yummain.py: Checks if we have read and write access in the current
	directory. If not, chdir to /. BZ 698795

2011-06-23  Zdeněk Pavlas <zpavlas@redhat.com>

	* output.py: Make the source xgettext-friendly.  xgettext ignores
	_(var), so localize literals instead.

2011-06-23  James Antill <james@and.org>

	* cli.py: Don't show -c warning, when we've changed it due to
	--installroot. BZ 707358.

2011-06-22  James Antill <james@and.org>

	* yum/rpmtrans.py:  Workaround rpm chroot insanity, so we can use sqlite from
	 unInstStop() cb.   So the problem is that AGAIN, when we log stuff
	 for installed rpms we do it in _instCloseFile, which is outside the
	 chroot ... but when we log stuff for erase rpms we do it in
	 _unInstStop, which is inside the chroot.  sqlite will open it's
	 journal file, which is it has a working path for from outside the
	 chroot ... being inside the chroot breaks this, which results in
	 complete failure of everything.   Fix is to move out of the chroot,
	 before we call the history/sqlite APIs, and then move back in again
	 ... basically what rpm does.

2011-06-22  James Antill <james@and.org>

	* ChangeLog, yum.spec, yum/__init__.py: Update ChangeLog/version for
	3.4.2 release.

2011-06-22  Zdeněk Pavlas <zpavlas@redhat.com>

	* output.py: Clean up the 'Transaction summary' report.

2011-06-21  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'yum-3_2_X' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum into yum-3_2_X  *
	'yum-3_2_X' of ssh://yum.baseurl.org/srv/projects/yum/git/yum:   Fix
	a leaked file descriptor

2011-06-21  Zdeněk Pavlas <zpavlas@redhat.com>

	* yum/__init__.py, yum/logginglevels.py: Fix a leaked file
	descriptor  'yum.log' is now closed when YumBase instance is GC'd.
	Resolves BZ 713924.

2011-06-21  Seth Vidal <skvidal@fedoraproject.org>

	* yum/packages.py: add check to make sure rest in the libc
	collapsing is not a single item list.

2011-06-20  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py: Uses ngettext to pluralise messages produced in
	__init__.py. BZ 552317

2011-06-20  Nick Jacek <njacek@redhat.com>

	* cli.py: Uses ngettext to pluralise messages produced by several
	functions in cli.py. BZ 552317

2011-06-20  Nick Jacek <njacek@redhat.com>

	* output.py: Changes the Transaction Summary so that the plurals are
	printed using ngettext. BZ 552317

2011-06-20  James Antill <james@and.org>

	* yum/__init__.py: Add message to history redo/undo for when nothing
	is done.

2011-06-16  James Antill <james@and.org>

	* yum/rpmtrans.py: Turn a name into a txmbr, if we can. Helps erase
	filelog/scriptout and history.

2011-06-16  James Antill <james@and.org>

	* yum/__init__.py: Add comments on TS_FAILED setting.

2011-06-16  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py, yum/rpmtrans.py: If verifyTransaction detects
	that an install or removal was not successful, it will now set the
	output_state of the pkg to TS_FAILED. BZ 661962  For the case of a
	removal, since the output_state will be TS_FAILED rather than just
	remaining TS_ERASE, yum will now correctly say that the removal
	failed instead of saying that the package was removed.

2011-06-16  James Antill <james@and.org>

	* yum/update_md.py: Add a __unicode__ to the UpdateNotice object, so
	the test code works.

2011-06-14  James Antill <james@and.org>

	* yum/__init__.py: Sort the packages, best we can easily, in
	searchPackageProvides().

2011-06-13  Nick Jacek <njacek@redhat.com>

	* yum/__init__.py, yum/depsolve.py:  Stops "Finished dependency resolution" from being printed multiple
	 times when --skip-broken is used. BZ 626569.

2011-06-13  Marcelo Moreira de Mello <mmello@redhat.com>

	* docs/yum.conf.5: - Fixed small typo at protected_multilib description on yum.conf man
	page  Hello,    This patches fixes a small type at
	protected_multilib description on yum.conf man page.   Cheers,
	Marcelo -- Marcelo Moreira de Mello  From: Marcelo Moreira de Mello
	<mmello@redhat.com> Date: Mon, 13 Jun 2011 16:10:07 -0300 Subject:
	[PATCH]  Fixed small typo at protected_multilib description at
	yum.conf man page

2011-06-13  James Antill <james@and.org>

	* yum/__init__.py: Show the correct "duplicate" package for search
	without --showduplicates.

2011-06-13  James Antill <james@and.org>

	* yum/packages.py: Always sort installed repos. after available
	repos.

2011-06-13  James Antill <james@and.org>

	* output.py: Show from_repo in matchcallback. Only print "Matched
	from" is there is more.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* docs/yum.8: Updated documentation for the new groups command.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* yumcommands.py: Remove the unused group command classes.

2011-06-08  Nick Jacek <njacek@redhat.com>

	* cli.py, yumcommands.py: Adds the Groups command, and removes
	individual command classes for grouplist etc. Adds the group summary
	function.

2011-06-06  Casey Jao <cjao@ugcs.caltech.edu>

	* test/complexremovetests.py: Add complex remove tests for new
	_remove_old_deps code.

2011-06-06  Casey Jao <cjao@ugcs.caltech.edu>

	* test/testbase.py: Make FakePackage use a simulated
	RPMDBAdditionalDataPackage for yumdb_info.


etc/yum.bash
# bash completion for yum

_yum_helper()
{
    local IFS=$'\n'
    COMPREPLY+=( $(
        /usr/share/yum-cli/completion-helper.py -d 0 -C "$@" 2>/dev/null ) )
}

_yum_list()
{
    # Fail fast for things that look like paths.
    [[ $2 == */* || $2 == [.~]* ]] && return
    _yum_helper list "$@"
}

# arguments:
#   1 = 1 or 0 to list enabled or disabled plugins
#   2 = current word to be completed
_yum_plugins()
{
    local val
    [[ $1 -eq 1 ]] && val='\(1\|yes\|true\|on\)' || val='\(0\|no\|false\|off\)'
    COMPREPLY+=( $( compgen -W '$( command grep -il "^\s*enabled\s*=\s*$val" \
        /etc/yum/pluginconf.d/*.conf 2>/dev/null \
        | sed -ne "s|^.*/\([^/]\{1,\}\)\.conf$|\1|p" )' -- "$2" ) )
}

# arguments:
#   1 = current word to be completed
_yum_binrpmfiles()
{
    COMPREPLY+=( $( compgen -f -o plusdirs -X '!*.rpm' -- "$1" ) )
    COMPREPLY=( $( compgen -W '"${COMPREPLY[@]}"' -X '*.src.rpm' ) )
    COMPREPLY=( $( compgen -W '"${COMPREPLY[@]}"' -X '*.nosrc.rpm' ) )
}

_yum_baseopts()
{
    local opts='--help --tolerant --cacheonly --config --randomwait
        --debuglevel --showduplicates --errorlevel --rpmverbosity --quiet
        --verbose --assumeyes --assumeno --version --installroot --enablerepo
        --disablerepo --exclude --disableexcludes --obsoletes --noplugins
        --nogpgcheck --skip-broken --color --releasever --setopt'
    [[ $COMP_LINE == *--noplugins* ]] || \
        opts+=" --disableplugin --enableplugin"
    printf %s "$opts"
}

_yum_transactions()
{
    COMPREPLY+=( $( compgen -W "$( $yum -d 0 -C history 2>/dev/null | \
        sed -ne 's/^[[:space:]]*\([0-9]\{1,\}\).*/\1/p' )" -- "$cur" ) )
}

_yum_atgroups()
{
    if [[ $1 == \@* ]]; then
        _yum_helper groups list all "${1:1}"
        COMPREPLY=( "${COMPREPLY[@]/#/@}" )
        return 0
    fi
    return 1
}

# arguments:
#   1 = current word to be completed
#   2 = previous word
# return 0 if no more completions should be sought, 1 otherwise
_yum_complete_baseopts()
{
    case $2 in

        -d|--debuglevel|-e|--errorlevel)
            COMPREPLY=( $( compgen -W '0 1 2 3 4 5 6 7 8 9 10' -- "$1" ) )
            return 0
            ;;

        --rpmverbosity)
            COMPREPLY=( $( compgen -W 'info critical emergency error warn
                debug' -- "$1" ) )
            return 0
            ;;

        -c|--config)
            COMPREPLY=( $( compgen -f -o plusdirs -X "!*.conf" -- "$1" ) )
            return 0
            ;;

        --installroot|--downloaddir)
            COMPREPLY=( $( compgen -d -- "$1" ) )
            return 0
            ;;

        --enablerepo)
            _yum_helper repolist disabled "$1"
            return 0
            ;;

        --disablerepo)
            _yum_helper repolist enabled "$1"
            return 0
            ;;

        --disableexcludes)
            _yum_helper repolist all "$1"
            COMPREPLY=( $( compgen -W '${COMPREPLY[@]} all main' -- "$1" ) )
            return 0
            ;;

        --enableplugin)
            _yum_plugins 0 "$1"
            return 0
            ;;

        --disableplugin)
            _yum_plugins 1 "$1"
            return 0
            ;;

        --color)
            COMPREPLY=( $( compgen -W 'always auto never' -- "$1" ) )
            return 0
            ;;

        -R|--randomwait|-x|--exclude|-h|--help|--version|--releasever|--cve|\
        --bz|--advisory|--tmprepo|--verify-filenames|--setopt)
            return 0
            ;;

        --download-order)
            COMPREPLY=( $( compgen -W 'default smallestfirst largestfirst' \
                -- "$1" ) )
            return 0
            ;;

        --override-protection)
            _yum_list installed "$1"
            return 0
            ;;

        --verify-configuration-files)
            COMPREPLY=( $( compgen -W '1 0' -- "$1" ) )
            return 0
            ;;
    esac

    return 1
}

_yum()
{
    COMPREPLY=()
    local yum=$1 cur=$2 prev=$3 words=("${COMP_WORDS[@]}")
    declare -F _get_comp_words_by_ref &>/dev/null && \
        _get_comp_words_by_ref -n = cur prev words

    # Commands offered as completions
    local cmds=( check check-update clean deplist distro-sync downgrade
        groups help history info install list makecache provides reinstall
        remove repolist search shell update upgrade version )

    local i c cmd subcmd
    for (( i=1; i < ${#words[@]}-1; i++ )) ; do
        [[ -n $cmd ]] && subcmd=${words[i]} && break
        # Recognize additional commands and aliases
        for c in ${cmds[@]} check-rpmdb distribution-synchronization erase \
            group groupinfo groupinstall grouplist groupremove groupupdate \
            grouperase localinstall localupdate whatprovides ; do
            [[ ${words[i]} == $c ]] && cmd=$c && break
        done
    done

    case $cmd in

        check|check-rpmdb)
            COMPREPLY=( $( compgen -W 'dependencies duplicates all' \
                -- "$cur" ) )
            return 0
            ;;

        check-update|makecache|provides|whatprovides|resolvedep|search)
            return 0
            ;;

        clean)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'expire-cache packages headers
                    metadata cache dbcache all' -- "$cur" ) )
            return 0
            ;;

        deplist)
            COMPREPLY=( $( compgen -f -o plusdirs -X '!*.[rs]pm' -- "$cur" ) )
            _yum_list all "$cur"
            return 0
            ;;

        distro-sync|distribution-synchronization)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'full different' -- "$cur" ) )
            _yum_list installed "$cur"
            return 0
            ;;

        downgrade|reinstall)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list installed "$cur"
            fi
            return 0
            ;;

        erase|remove)
            _yum_atgroups "$cur" || _yum_list installed "$cur"
            return 0
            ;;

        group*)
            if [[ $cmd == groups || $cmd == group && $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'info install list remove summary' \
                    -- "$cur" ) )
            else
                _yum_helper groups list all "$cur"
            fi
            return 0
            ;;

        help)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W '${cmds[@]}' -- "$cur" ) )
            return 0
            ;;

        history)
            if [[ $prev == $cmd ]] ; then
                COMPREPLY=( $( compgen -W 'info list packages-list
                    packages-info summary addon-info redo undo rollback new
                    sync stats' -- "$cur" ) )
                return 0
            fi
            case $subcmd in
                undo|repeat|addon|addon-info|rollback)
                    if [[ $prev == $subcmd ]]; then
                        COMPREPLY=( $( compgen -W "last" -- "$cur" ) )
                        _yum_transactions
                    fi
                    ;;
                redo)
                    case $prev in
                        redo)
                            COMPREPLY=( $( compgen -W "force-reinstall
                                force-remove last" -- "$cur" ) )
                            _yum_transactions
                            ;;
                        reinstall|force-reinstall|remove|force-remove)
                            COMPREPLY=( $( compgen -W "last" -- "$cur" ) )
                            _yum_transactions
                            ;;
                    esac
                    ;;
                package-list|pkg|pkgs|pkg-list|pkgs-list|package|packages|\
                packages-list|pkg-info|pkgs-info|package-info|packages-info)
                    _yum_list available "$cur"
                    ;;
                info|list|summary)
                    if [[ $subcmd != info ]] ; then
                        COMPREPLY=( $( compgen -W "all" -- "$cur" ) )
                        [[ $cur != all ]] && _yum_list available "$cur"
                    else
                        _yum_list available "$cur"
                    fi
                    _yum_transactions
                    ;;
                sync|synchronize)
                    _yum_list installed "$cur"
                    ;;
            esac
            return 0
            ;;

        info)
            _yum_list all "$cur"
            return 0
            ;;

        install)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list available "$cur"
            fi
            return 0
            ;;

        list)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all available updates installed
                    extras obsoletes recent' -- "$cur" ) )
            return 0
            ;;

        localinstall|localupdate)
            _yum_binrpmfiles "$cur"
            return 0
            ;;

        repolist)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all enabled disabled' -- "$cur" ) )
            return 0
            ;;

        shell)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -f -o plusdirs -- "$cur" ) )
            return 0
            ;;

        update|upgrade)
            if ! _yum_atgroups "$cur" ; then
                _yum_binrpmfiles "$cur"
                _yum_list updates "$cur"
            fi
            return 0
            ;;
        version)
            [[ $prev == $cmd ]] && \
                COMPREPLY=( $( compgen -W 'all installed available nogroups
                    grouplist groupinfo' -- "$cur" ) )
            return 0
            ;;
    esac

    local split=false
    declare -F _split_longopt &>/dev/null && _split_longopt && split=true

    _yum_complete_baseopts "$cur" "$prev" && return 0

    $split && return 0

    if [[ $cur == -* ]] ; then
        COMPREPLY=( $( compgen -W '$( _yum_baseopts )' -- "$cur" ) )
        return 0
    fi
    COMPREPLY=( $( compgen -W '${cmds[@]}' -- "$cur" ) )
} &&
complete -F _yum -o filenames yum yummain.py

# Local variables:
# mode: shell-script
# sh-basic-offset: 4
# sh-indent-comment: t
# indent-tabs-mode: nil
# End:
# ex: ts=4 sw=4 et filetype=sh
